### PR TITLE
store/tikv: add GC worker.

### PR DESCRIPTION
**Pull request by disksing**

### CODE REVIEW

**Comment by disksing** [comment]
PTAL @siddontang @shenli

File: store/tikv/gc_worker.go [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+		if len(key) == 0 {
+			break
+		}
+	}
+	log.Infof("[gc worker] %s finish gc, safePoint: %v, regions: %v, cost time: %s", w.uuid, safePoint, regions, time.Now().Sub(startTime))
+	return nil
+}
+
+func (w *GCWorker) checkLeader() (bool, error) {
+	leader, err := w.loadValueFromSysTable(gcLeaderUUIDKey)
```
Comment (shenli) [inline store/tikv/gc_worker.go]: Should we use a transaction to check leader?

File: store/tikv/gc_worker.go [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+	session, err := tidb.CreateSession(store)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	ver, err := store.CurrentVersion()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	worker := &GCWorker{
+		uuid:    uuid.NewV4().String(),
```
Comment (shenli) [inline store/tikv/gc_worker.go]: Why not use ver as worker id?

File: store/tikv/gc_worker.go (hunk 1 above)
Comment (disksing) [inline store/tikv/gc_worker.go]: @shenli I also thought use a transaction would be better, but it seems that `tidb.Session` does not provide a 'NewTxn' method...
btw, it won't matter even if 2 workers unfortunately becomes leader in the same time.

**Comment by shenli** [comment]
LGTM

**Comment by shenli** [comment]
@coocood @siddontang @zimulala PTAL

File: store/tikv/txn_committer.go [hunk 3]
```diff
@@ -319,6 +319,11 @@ func (c *txnCommitter) cleanupKeys(bo *Backoffer, keys [][]byte) error {
 	return c.iterKeys(bo, keys, c.cleanupSingleRegion, c.keySize, false)
 }
 
+// The max time a Txn may use (in ms) from its startTS to commitTS.
+// We use it to guarantee GC worker will not influence any active txn. The value
+// should be less than `gcRunInterval`.
+const maxTxnTimeUse = 590000
```
Comment (zimulala) [inline store/tikv/txn_committer.go]: Using 590 * time.Second is better.

File: store/tikv/gc_worker.go [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+			}
+			ver, err := w.store.CurrentVersion()
+			if err != nil {
+				log.Errorf("[gc worker] failed get current version: %v", err)
+				break
+			}
+			safePoint = ver.Ver
+		case <-w.quit:
+			log.Infof("[gc worker] (%s) quit.", w.uuid)
+			break
```
Comment (zimulala) [inline store/tikv/gc_worker.go]: If you want quit the start function, break can't do it. Break only breaks select here.

File: store/tikv/txn_committer.go (hunk 3 above)
Comment (shenli) [inline store/tikv/txn_committer.go]: 590 * time.Second = 590 * 1000 * 1000 * 1000 != 590000

File: store/tikv/gc_worker.go [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+			return errors.Trace(err)
+		}
+		if regionErr := resp.GetRegionError(); regionErr != nil {
+			err = bo.Backoff(boRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
+		gcResp := resp.GetCmdGcResp()
```
Comment (disksing) [inline store/tikv/gc_worker.go]: It's defined as `cmd_gc_resp` in .proto file and generated by protoc. If we want 'CmdGCResp' then we will have to define `cmd_gC_resp` in .proto...

**Comment by shenli** [comment]
@zimulala PTAL

File: store/tikv/gc_worker.go [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+			return errors.Errorf("unexpected scanlock error: %s", locksResp)
+		}
+		locksInfo := locksResp.GetLocks()
+		locks := make([]*Lock, len(locksInfo))
+		for i := range locksInfo {
+			locks[i] = newLock(locksInfo[i])
+		}
+		ok, err1 := w.store.lockResolver.ResolveLocks(bo, locks)
+		if err1 != nil {
+			return errors.Trace(err)
```
Comment (zimulala) [inline store/tikv/gc_worker.go]: errors.Trace(err) -> errors.Trace(err1)?

File: store/tikv/gc_worker.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+			return
+		}
+	}
+}
+
+func (w *GCWorker) runGCJob(safePoint uint64) {
+	log.Infof("[gc worker] %s starts GC job, safePoint: %v", w.uuid, safePoint)
+	err := w.resolveLocks(safePoint)
+	if err != nil {
+		w.done <- errors.Trace(err)
```
Comment (zimulala) [inline store/tikv/gc_worker.go]: If resolveLocks failed then it doesn't run doGC. So it needs a `return`.

File: store/tikv/gc_worker.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if regionErr := resp.GetRegionError(); regionErr != nil {
+			err = bo.Backoff(boRegionMiss, errors.New(regionErr.String()))
+			if err != nil {
+				return errors.Trace(err)
+			}
+			continue
+		}
```
Comment (zimulala) [inline store/tikv/gc_worker.go]: Line137-157 and line207-227 code are look like duplicated. Can they make a common method?

Comment (disksing) [inline store/tikv/gc_worker.go]: @zimulala I prefer not to do it. The extracted method would be `Foo(bo, req, key) (resp, region, shouldContinue, error)`, which may make it more complicated.

File: store/tikv/gc_worker.go [hunk 9]
```diff
@@ -0,0 +1,332 @@
...
+			return false, errors.Trace(err)
+		}
+		return true, nil
+	}
+	w.session.Execute("ROLLBACK")
+	return false, nil
+}
+
+func (w *GCWorker) updateLease() error {
+	lease := time.Now().Add(gcWorkerTickInterval * 2).Format(gcLeaderLeaseFormat)
```
Comment (zimulala) [inline store/tikv/gc_worker.go]: gcWorkerTickInterval * 2 -> gcWorkerLease ?
gcWorkerLease  hasn't been used.

**Comment by zimulala** [comment]
Rest LGTM

**Pull request merged**